lib/parser/message.cc

```diff
@@ -33,29 +33,82 @@ std::ostream &operator<<(std::ostream &o, const MessageFixedText &t) {
   return o;
 }
 
+static bool NeedsSpecialFormatting(const char *p) {
+  while (*p != '\0') {
+    if (*p++ == '%') {
+      if (*p == 'S' || *p == 'B') {
+        return true;
+      } else if (*p != '%' && *p != 's' && *p != 'd' &&
+          !(p[1] == 'd' && (*p == 'z' || *p == 'j'))) {
+        return false;
+      }
+    }
+  }
+  return false;
+}
+
+static std::string SpecialFormatting(const char *p, va_list &ap) {
+  std::string result;
+  while (*p != '\0') {
+    if (*p != '%') {
+      result += *p++;
+    } else {
+      ++p;
+      switch (*p++) {
+      case '%': result += '%'; break;
+      case 's': result += va_arg(ap, const char *); break;
+      case 'd': result += std::to_string(va_arg(ap, int)); break;
+      case 'S': result += va_arg(ap, std::string); break;
+      case 'B': result += va_arg(ap, CharBlock).ToString(); break;
+      default:
+        if (*p == 'd') {
```

Comment:
Could other cases such as 'u' occur here?

Reply:
Sure.  Any formatting accepted by the underlying `vsnprintf()` can appear, and if it's not handled by the cases in this code, it'll end up being handled in `vsnprintf()` below.

Reply:
I'm puzzled as to why %jd is different than $ju

Reply:
The difference is that once we hit a format that this code doesn't special-case, it goes off to `vsnprintf()` to deal with the rest of the formatting, so any instances of the special `%B` and `%S` formats after that point won't be expanded.  So there's some special casing here of those format codes that might appear before a `%B` or `%S` so that we don't have to call `vsnprintf()` for them.  Is there an instance of `%ju` formatting anywhere?

Reply:
I have a %ju in new code.   There might also be future instances.